openpgp/packet/aead_encrypted.go

```diff
@@ -0,0 +1,401 @@
+// Copyright (C) 2019 ProtonTech AG
+
+package packet
+
+import (
+	"bytes"
+	"crypto/cipher"
+	"crypto/rand"
+	"encoding/binary"
+	"io"
+
+	"golang.org/x/crypto/openpgp/errors"
+)
+
+// AEADEncrypted represents an AEAD Encrypted Packet (tag 20, RFC4880bis-5.16).
+type AEADEncrypted struct {
+	cipher        CipherFunction
+	mode          AEADMode
+	chunkSizeByte byte
+	Contents      io.Reader // Encrypted chunks and tags
+	initialNonce  []byte    // Referred to as IV in RFC4880-bis
+}
+
+// An AEAD opener/sealer, its configuration, and data for en/decryption.
+type aeadCrypter struct {
+	aead           cipher.AEAD
+	config         *AEADConfig
+	initialNonce   []byte
+	associatedData []byte // Chunk-independent associated data
+	chunkIndex     []byte // Chunk counter
+	bytesProcessed int    // Amount of (plain/cipher)-text bytes read/written
+	cache          bytes.Buffer
+}
+
+// aeadEncrypter encrypts and writes bytes. It encrypts when necessary according
+// to the AEAD block size, and caches the extra encrypted bytes for next write.
+type aeadEncrypter struct {
+	aeadCrypter                // Embedded plaintext sealer
+	writer      io.WriteCloser // 'writer' is a partialLengthWriter
+}
+
+// aeadDecrypter reads and decrypts bytes. It caches extra decrypted bytes when
+// necessary, similar to aeadEncrypter.
+type aeadDecrypter struct {
+	aeadCrypter           // Embedded ciphertext opener
+	reader      io.Reader // 'reader' is a partialLengthReader
+	peekedBytes []byte    // Used to detect last chunk
+}
+
+func (ae *AEADEncrypted) parse(buf io.Reader) error {
+	headerData := make([]byte, 4)
+	if n, err := buf.Read(headerData); err != nil || n < 4 {
+		return errors.AEADError("could not read aead header")
+	}
+	// Read initial nonce
+	mode := AEADMode(headerData[2])
+	nonceLen := mode.NonceLength()
+	initialNonce := make([]byte, nonceLen)
+	if n, err := buf.Read(initialNonce); err != nil || n < nonceLen {
+		return err
+	}
+	ae.Contents = buf
+	ae.initialNonce = initialNonce
+	ae.cipher = CipherFunction(headerData[1])
```

Comment:
Check if `CipherFunction(headerData[1])` is supported. `symmetrically_encrypted` hacks this with `KeySize`.

Reply:
Yeah, perhaps it would be cleaner to do something like:
```go
if ae.cipher, ok = algorithm.CipherById[headerData[1]]; !ok {
	return errors.UnsupportedError("unknown cipher: " + strconv.Itoa(int(headerData[1])))
}
```
Though.. then you're using internal stuff where you weren't previously. (Though that also doesn't seem terrible, and I'm not 100% sure of the added value of having an internal and external type for the same thing.) If you define `KeySize` to return 0 for unsupported ciphers, then it also doesn't seem too terrible to use it for that. Anyway, I think both are fine.

Reply:
Yes, not too terrible, but it's still kind of silly. The use of that map doesn't convince much (it's only used once, in `public_key.go`). 
 This `KeySize` practice is done in more places, for me the code really deserves an `IsSupported` function.

Reply:
So - to me, having
```go
ae.cipher = CipherFunction(headerData[1])
if !ae.cipher.IsSupported() { ... }
```
doesn't make the most sense, because why can you construct a CipherFunction that doesn't exist? You're relying on the programmer to check, instead of letting the type system work for you.
So in my mind, a better API would be something where that's impossible in the first place, e.g. something like
```go
ae.cipher = IdToCipherFunction(headerData[1]) // throws if it doesn't exist
```
but that's not how Go works, instead, you have to write
```go
if ae.cipher, err := IdToCipherFunction(headerData[1]); err { ... }
```
which is not too different from
```go
if ae.cipher, ok := CipherById[headerData[1]]; !ok { ... }
```
so that's why that version makes sense to me.

Reply:
Well, `HashIdToHash` just uses the `HashById` map, so it's basically the code you suggest, disguised.
But Okay, `CipherIdToCipher` it is.